### ISSUE TITLE
[RFC FS-1003 Discussion] nameof Operator

### ISSUE THREAD

**Issue by enricosada** [fsharp/fslang-design#48]
This issue is used to track discussions of [F# RFC FS-1003 - "nameof Operator"](https://github.com/fsharp/fslang-design/blob/master/FSharp-5.0/FS-1003-nameof-operator.md). Please discuss in thread below (if necessary)

**Comment by dsyme** [comment]
I added some comments to the proposed PR here: https://github.com/Microsoft/visualfsharp/pull/908#discussion_r51915192

I'm concerned about the implementation strategy and think it should probably be done earlier in the type checker.  

Some clarifications to the spec are needed

- Can nameof be used on ambiguous overloaded methods?
- Can nameof be used on properties that only have setters?
- Can nameof be used on constructors?

**Comment by thinkbeforecoding** [comment]
Since nameof is just the name, ambiguous overloaded methods still all have the same name, so it should work.
Same thing for properties with only setters.
For ctors, the question is.. What's the name of a constructor... If it's the same as the type, we can just use that.

**Comment by dsyme** [comment]
@thinkbeforecoding I believe ``nameof(C.M)`` will raise an ambiguous-overload type checking error in the current implementation if C.M is overloaded?

**Comment by forki** [comment]
>> I'm concerned about the implementation strategy and think it should probably be done earlier in the type checker.

@dsyme where should I try to hook it in?

**Comment by dsyme** [comment]
I think it's captured by this comment: 

> I believe we should reconsider the implementation of the feature and intercept nameof during type checking. Instead of checking the argument we would require that it is a long identifier and do unqualified expression name resolution on that identifier to a symbol, and then just take the name of the symbol. I believe that is more in the spirit of how the feature is intended to work.

So there would be a special case in the type checker

**Comment by forki** [comment]
Yes I got that. I was wondering if we already have a place where we pattern
match on the tast and rewrite it.

**Comment by dsyme** [comment]
Perhaps the closest is in ``TcFunctionApplicationThen`` where we look for ``seq { ... }`` before we check the argument.  You could modify that to look for a special ``nameof`` value, and then apply completely different rules/restrictions for the processing of the argument.

There are also similarities with the processing of the ``reraise`` special value, which is only allowed at certain positions.  But that has no special type-checking rules and its argument is processed as a normal argument.

**Comment by smoothdeveloper** [comment]
I don't see any mention of `[<CompiledName("name")>]` and I think this is an area which could get users in "principle of least surprise failed" (either way this would be handled I guess).

What is the expected behaviour in interaction with that attribute?

**Comment by thinkbeforecoding** [comment]
Interesting question. 
The feature seems quite private to F# and does not seem to be meant for reflection purpose - at least not to get a method info from name....
So I'd say it should return the F# name, not the compiled name.

**Comment by Thorium** [comment]
Main purpose of nameof  is probably some logging purposes, and it would be useful on some async scenarios when the stacktrace won't tell you all and you can't reraise. I would also vote the name in F# source code.

**Comment by Rickasaurus** [comment]
> Main purpose of nameof is probably some logging purposes, and it would be useful on some async scenarios when the stacktrace won't tell you all and you can't reraise. I would also vote the name in F# source code.

I want nameof mostly for some amount of type safety in my reflection code. When the name of a type is changed I want to be able to get an error (or gui rename-update) for that change.

**Comment by dsyme** [comment]
New implementation at https://github.com/Microsoft/visualfsharp/pull/2290

**Comment by smoothdeveloper** [comment]
@Rickasaurus comment makes me think we could also have `exprTypeOf` operator which returns a `System.Type` in reflection scenarios, does that sound like "fslang-design new issue" worthy?

**Comment by abelbraaksma** [comment]
I had some trouble understanding the whole requirements list on [RFC FS-1003](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1003-nameof-operator.md), would it be ok if I pulled it and add one or two examples to each bullet point for clarity? 

I also noticed two other things that are either missing or deliberately not included:

* Works with both fully qualified names, `global::` and unqualified names, provided they are in scope (i.e. `nameof String` and `nameof System.String` _(yes, this seems obvious, but implementation details are not part of the RFC and it could've been implemented at tokenization phase)_.
* Works with constants (i.e., it should be allowed to create a literal)

Also, I was wondering about the first couple of bullet points. It seems to me this should work as a keyword (like `if` or `let`), after all it is resolved at compile time, but some examples suggest it can be used assigned to a function, piped to etc. I don't necessarily object to that, but is that the current train of thought of how this is supposed to be implemented, or is discussion still open?

**Comment by dzmitry-lahoda** [comment]
C# got some nice "statically" typed patterns with this. Hope F# would get that soon.

**Comment by vasily-kirichenko** [comment]
@dzmitry-lahoda I don’t think anyone is gonna finish it, the last attempt was mine, rebased and fixed @forki’s implementation, the last thing was left to done is making it work it on overloads, Don promised to help, but only promised, as it happens more and more these days.

**Comment by dzmitry-lahoda** [comment]
>We will enable and encourage strong community participation in F# by continuing to build the necessary infrastructure and tooling to complement community contributions. We will make F# the best-tooled functional language on the market, by improving the language and tooling experience, removing road blocks for contributions, and addressing pain points to `narrow the experience gap with C# and VB`. As new language features appear in C#, we will ensure that they also interoperate well with F#. F# will continue to target platforms that are important to its community. [1]


[1]: https://blogs.msdn.microsoft.com/dotnet/2017/02/01/the-net-language-strategy/

**Comment by abelbraaksma** [comment]
> Don promised to help, but only promised, as it happens more and more these days.

of I see the amount of work on his plate, that's not too surprising, there are so many issues, it's hard to keep up, and also keep improving the language with new features. I highly respect how much he and others of the team are participating, and surely, some issues and up under the rug, though I doubt that's intentional in any way.

Also, I don't think you we should strive for 100 percent parity, not all C# features are equally well suited for F#, but that's another discussion entirely.

**Comment by dsyme** [comment]
The revamped implementation of this feature is at https://github.com/Microsoft/visualfsharp/pull/6325

**Comment by jwosty** [comment]
For reference, this is being continued in: [dotnet/fsharp#6809](https://github.com/dotnet/fsharp/pull/6809)

**Comment by cartermp** [comment]
Closing as this is in and shipping

**Comment by dsyme** [comment]
Reopened as the feature is in preview

**Comment by cartermp** [comment]
I think something that could be specified in the RFC is some formalism around what `'T` actually is for the `nameof` function. For example, we had two people already mention that taking the name of an instance member, without the instance, should be allowed. But that's not a valid expression in the traditional sense. Yet it makes perfect sense for `nameof`. So that implies a new kind of expression, similar to how C# does it by defining a `named-expression`.

**Comment by dsyme** [comment]
> I think something that could be specified in the RFC is some formalism around what 'T actually is for the nameof function. 

Yes, I agree, it's currently syntactically an expression but is processed as a name resolution item (with slightly varied resolution rules).

**Comment by cartermp** [comment]
Added: https://github.com/fsharp/fslang-design/blob/master/preview/FS-1003-nameof-operator.md#alternative-formalize-the-notion-of-a-named-entity

**Comment by Happypig375** [comment]
Please implement for literal definitions.
```fs
let rec [<Literal>] SomeBindedPropertyName = nameof SomeBindedPropertyName
// Error FS0837 This is not a valid constant expression
// Error FS0267 This is not a valid constant expression or custom attribute value
// Error FS0039 The value or constructor 'SomeBindedPropertyName' is not defined.
```
Compare:
```fs
let rec SomeBindedPropertyName = nameof SomeBindedPropertyName // OK for F# non-literals
```
```cs
const string SomeBindedPropertyName = nameof(SomeBindedPropertyName); // OK in C#
```

**Comment by abelbraaksma** [comment]
I thought the rfc mentioned somewhere that this operator was allowed in literals, or at least that it was planned.

**Comment by Happypig375** [comment]
Use in attributes is permitted but not in literals.

**Comment by abelbraaksma** [comment]
I see, you are right, there indeed isn't any mention of literals or constants in the [RFC text](https://github.com/fsharp/fslang-design/blob/master/preview/FS-1003-nameof-operator.md). In fact, there is only one mention of a "can be used _in_", which could perhaps be expanded with some locations where it can, or cannot be used.

Though, arguably, your example uses `SomeBindedPropertyName`, which to me suggests that it points to an instance property. At the moment, an instance of a class is needed for instance members (as opposed to static members). Same is true for overloads. Obviously, that can't work in attributes, nor in literals.

**Comment by Happypig375** [comment]
This won't compile either:
```fs
let rec [<Literal>] X = nameof X
```

**Comment by Happypig375** [comment]
`member val`s do not have the ability to refer to themselves. Therefore, `nameof` on themselves are not permitted.
```fs
type X() as this =
    inherit Dictionary<string, string>()
    member val XValue = this.TryGetValue("XValue") |> function true, x -> ValueSome x | false, _ -> ValueNone // OK
    member val YValue = this.TryGetValue(nameof YValue) |> function true, y -> ValueSome y | false, _ -> ValueNone // FS0039 The value or constructor 'YValue' is not defined.
    member val rec ZValue = this.TryGetValue(nameof ZValue) |> function true, z -> ValueSome z | false, _ -> ValueNone // FS0010 Unexpected keyword 'rec' in member definition. Expected identifier or other token.
```
Please enable this.

**Comment by cartermp** [comment]
@Happypig375 https://github.com/fsharp/fslang-design/pull/403

**Comment by dsyme** [comment]
My preferred resolution of these:

> `nameof` on an operator gives the compiled name of the operator

Resolve as by design (I'm really not fussed by this nit, and it is really expensive to resolve - e.g. if the user explicitly uses `op_Addition` do they get `op_Addition`?)

> [ ] Resolving instance members requires an artificial instance object.

Resolve as by design (I'm really not fussed by this nit)

> [ ] `nameof` may not be used with the name of a generic type parameter.

Resolve as by design (I'm really not fussed by this nit, and the decision was by design because expression syntax doesn't include type variables)

> [ ] Resolving overloaded members requires a type instantiation.

Resolve as by design (I'm really not fussed by this nit)

> [ ] Using `nameof` in a literal, e.g.., `let rec [<Literal>] Foo = nameof Foo`

Fix this for sure

> [ ] Sepcifying `nameof` on a member val

Have you tried

    nameof this.YValue

(I haven't double checked as would need to compile the branch but I think it will work)

- [ ] Specifying `nameof` on  a value without needing to use `rec`, e.g., `let x = nameof x`

This is certainly by design, why would you be able to use `nameof` when `x` is not in scope?

**Comment by cartermp** [comment]
I think if we go with the current design where the value is a normal value, then (7) makes sense to resolve as you've put it. But Given that there's already a desire to make it a _named-expr_ akin to C# then perhaps different resolution rules can apply, much like how in C# you can take the name of something by fully qualifying it even though it's not a static member. I'm personally fine with requiring `rec`, but worth thinking about.

**Comment by dsyme** [comment]
I feel the guiding principle for `nameof(x)` is that it roughly corresponds to `statichandle(x).Name`, for some general `statichandle(x)` construct that returns whatever reflection handle is available for `x`.  This construct doesn't actually exist but it could - FSharp.Core for example has a `methodhandle(x)` construct.

So I guess the question is "what's the natural F# way to get a static handle for the thing being defined in `let x = ...`".  I don't have an immediate answer for that as it's just not a normal thing to be able to do in an ML-family language.  Simply referring to the thing via `x` doesn't seem at all right as there might be another `x` in scope, e.g.

```
type x = A | B

let f x = 
    let x, y, z = nameof(x), statichandle(x), typeof(x)
    (x, y, z)
```
Here each `x` potentially resolves to a different semantic thing. Now, for `nameof(x)` no one cares if you're logically speaking referring to a different semantic thing because they have the same name.  But I'm just queasy with having random rules for `nameof`, it will just be a bug farm and full of whacky corner cases.

So if `x` is not in scope the whole language feature just to feel iffy - a somewhat random collection of complex new resolution rules, for little clarifying purpose, and not orthogonal to anything else in the language.  I'm somewhat OK with this approach if it focuses very much on software-engineering utility, which is the only rationale for this whole feature - but I feel we've done that with the feature already and it captures the 99% of utility.

**Comment by abelbraaksma** [comment]
On (2) and (4), I don't know what the complexity is of resolving this, but a common use case for `nameof` is in logging and exception message generation.

I can think of numerous issues with requiring an instance, but to name the two biggest elephants in the room: side effects of constructors, and impossibility of using `nameof` from within a constructor (on instance members). Imo these are severe limitations on what ought to be only a compile time feature.

I don't know if the artifical instance is optimized away, but if it is, that would give other problems, for instance with predictability of code execution or the 'whatifs' when the instance is reused.

It also limits its use cases, i.e. instances cannot be used in attributes or literals, and the production of a `nameof` expression should yield a constant. Therefore, its expression must be a constant expression. Requiring an instance negates that. 

Finally, I think it's just hard to explain that for a supposedly compile time feature you need an instance.

Resolving this might also resolve (4), which aids the usefulness.

**Comment by cartermp** [comment]
> I'm somewhat OK with this approach if it focuses very much on software-engineering utility, which is the only rationale for this whole feature

This is correct. It's a compile-time only feature that aids primarily in the removal of magic strings in codebases, making things easier to refactor over periods of time. It's already not a normal function (you can't just call it with an older compiler, you can't use it as a first-class value), so I think we've sort of crossed the rubicon on the issue of not being orthogonal to the rest of the language.

Needing to add type annotations or construct an ephemeral object just feels strange for this kind of feature. If it were a normal function that at _runtime_ gives you a string, sure. But that's [clearly not the case here](https://github.com/dotnet/fsharp/blob/7adaacf150dd79f072efe42d43168c9cd6edbced/src/fsharp/FSharp.Core/prim-types.fs#L4144).

**Comment by dsyme** [comment]
> Finally, I think it's just hard to explain that for a supposedly compile time feature you need an instance.

Note you can use `Unchecked.defaultof<C>.M` for the instance

The F# name resolution rules just make `C.M` a static method and `c.M` an instance method (for instance `c`), and there are subtle differences in the name resolution for each.   Is the proposal that `C.M` be usable for nameof as well?  I mean, it's possible I think, but just odd to have whole new sections of the spec and implementation for this one construct.

Anyway yes, for (2) and (4) I'm ok with making the type annotations optional and allowing C.M for instance methods.  I'm no great fan but hey...

> so I think we've sort of crossed the rubicon on the issue of not being orthogonal to the rest of the language.

To my knowledge we've not crossed the line where it is unaligned with the resolution rules for some future `statichandleof` or `propertyof` or public `methodof` etc.   Note that all of these are genuinely useful for removing magic logic from reflection programming, like `typeof`.

The main thing here is (4), i.e. `methodof` requires a type instantiation to pinpoint an exact method overload.

Perhaps we should aim for `nameof` and other future `xyzof` to align apart from (4).

**Comment by abelbraaksma** [comment]
> Note you can use `Unchecked.defaultof<C>.M` for the instance

In some scenarios, yes (though I didn't know that a null reference would work here, it is good to know). But an important use-case for `nameof` is in constants and in attributes that require a string that points to a member. For that to work, the `expr` in `nameof (expr)`, must itself be expressible as a constant expression for instance members.

With other use-cases, like for making code with `ArgumentNullException` etc more maintainable, this is less of a concern I guess. 

> Anyway yes, for (2) and (4) I'm ok with making the type annotations optional and allowing C.M for instance methods. I'm no great fan but hey...

I understand the reluctance, but C# [does it the same way](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/nameof), which may ease the transition between the languages.

> The main thing here is (4), i.e. methodof requires a type instantiation to pinpoint an exact method overload.

I don't understand this part, if there's a method group (of overloads) I assume it doesn't matter whether you pick the first, the last or anything in-between, they will all have the same name anyway, or is this is a detail that is simply difficult to work-around given the current state of internal compiler library methods?

**Comment by cartermp** [comment]
Yeah, I don't quite understand the issue with overloads here. C# doesn't do any overload resolution, since each named method is the same. I brought this up in the original PR a while back. I certainly wouldn't expect a different name based on overloading.

> Note that all of these are genuinely useful for removing magic logic from reflection programming

I imagine so. And we may need to consider those cases differently should they be on the table for a future F# release. But I don't think this quite applies to `nameof`. Although `nameof` is helpful when writing reflection code, the primary utility is as a software engineering tool that lets you rename constructs over time without fear that you're logging a name of something incorrectly. It feels kind of unique in that sense, hence the alternative proposal that we consider a similar design to C# wherein the `expr` is really a `named-expr` with slightly different resolution rules. It certainly feels more natural when using it in practice to do it that way.

**Comment by dsyme** [comment]
> Although nameof is helpful when writing reflection code, the primary utility is as a software engineering tool that lets you rename constructs over time without fear that you're logging a name of something incorrectly. 

Right, but that's exactly the point - we must have accuracy in knowing which symbol things resolve to in order to implement renaming.  So for example consider renaming the different 'x' here:

```
let f  x =
    let x = nameof x
    let x = nameof x
    x
```
When you rename the `x` in the first `let x` does the `nameof(x)` rename?  The suggestion to make `x` accessible out of scope would make the renaming very strange here.

The same applies to method overloading.  In the VF# tools, renaming is based on a specific overload, e.g.
```
type C() = 
    member c.M1() = 1
    member c.M1(x: int) = 1

module Say =
    let c = C()

    let y = c.M1()
    let z = c.M1(3)
    0

```
When you rename the first `M1` only that specific overload is renamed.  So for 
```
    let n = nameof(c.M1)
```
do we rename or not? There is no right answer anymore....

**Comment by dsyme** [comment]
> I don't understand this part, if there's a method group (of overloads) I assume it doesn't matter whether you pick the first, the last or anything in-between, they will all have the same name anyway

That's correct for `nameof` though not for `methodof`

**Comment by Happypig375** [comment]
C# supports this, therefore it has a resolution.

**Comment by abelbraaksma** [comment]
> do we rename or not? There is no right answer anymore....

@dsyme That is true. But programmers are not entirely oblivious of this either. If they want proper renaming down to the member of the method group of overloads, they can use an annotated version with an instance. When they use an ambiguous match by using a constant expression, either because they have to, or for esthetic reasons, the compiler will pick the first from the group.

Since renaming of something in a group usually (but arguably not always) will end up renaming the whole group, this is not a problem. The compiler an pick the last in declaration order, as if it's an overloaded let binding. 

I can live with that limitation. It's probably the same as for C# anyway. There is no complete solution, so we should leave it to the programmer to use it properly, but we shouldn't limit the use artificially.

> C# supports this, therefore it has a resolution.

@Happypig375 That is not a logical conclusion, even though it'd be true in many cases. And in this case it isn't. They didn't resolve it, but allowed it. In C#, if you use an overload, and you rename any member, the member in `nameof`  is not renamed. Only after you rename all members, the member in `nameof` is also renamed.

We could take the same route as C#: renaming of the argument in `nameof` is only carried out if there're no overloads anymore. 

> The suggestion to make x accessible out of scope would make the renaming very strange here.

I agree, let's use the same scoping rules as for functions as much as possible. `x` isn't in scope of itself, so be it ;). 

Btw, what about aliases? These aren't allowed in C# (but so aren't operators), but I don't see why we wouldn't.

**Comment by Happypig375** [comment]
>In C#, if you use an overload, and you rename any member, the member in nameof is not renamed. Only after you rename all members, the member in nameof is also renamed.

Sounds like a resolution to me. It's not like this behavior is going to change anytime soon.

**Comment by abelbraaksma** [comment]
My point was, it doesn't resolve the fact that `nameof` is supposed to protect against renames, in the case of overloads, it doesn't. And while I can confirm the behaviour, the C# language spec doesn't mention this.

Regardless, I think it makes sense to move in the same direction, unless that requires a change that isn't worth the effort. In the latter case, I suggest we make it an error.

**Comment by cartermp** [comment]
@dsyme renaming for overloads in C# is handled as you would expect: the entire set is renamed if initiated from the named expression. If you rename an overload itself, then the named expression within the `nameof` is not renamed. And if there is only one corresponding method then it's fine. This is likely by design. So I don't think this is intractable, we just need to decide if we prefer one of the two scenarios more:

* You can take a name of an overloaded method without having to annotate it to say which overload you mean
* You can rename any overload and all symbols in a `nameof` will also be renamed

I would prefer the first scenario, since people generally write code a lot more than they refactor it.

**Comment by abelbraaksma** [comment]
> You can take a name of an overloaded method without having to annotate it to say which overload you mean

I think that ought to be the minimum (with the caveat that it may be somewhat beneficial to have the option to use annotations for these fringe cases that this is actually useful, or is it?)

> You can rename any overload and all symbols in a nameof will also be renamed

Wouldn't be my preference either, as it is opposite to C# behavior. I think it somehow makes sense: a method inside `nameof` belongs to a method group, then renaming one in that group effectively takes it out. If the nameof expression would be renamed then, this would take it out of the group as well.

It makes more sense to leave it in the group until the group is not a group anymore. 

I think the inverse makes sense as well, rename the expression in `nameof` renames the whole group, by reasoning that the one nameof expression represents the whole group.

**Comment by dsyme** [comment]
@cartermp Yes, that's a reasonable resolution for renaming. Quite tricky to implement I think, I'm not sure how critical it is as long as there is a design that makes sense in principle.

> You can take a name of an overloaded method without having to annotate it to say which overload you mean

Yes, I'm ok with this

**Comment by dharmaturtle** [comment]
Calling `nameof` on instance members yields an error message:

>Using the 'nameof' operator as a first-class function value is not permitted.

Can the error message just say something like "Instance members are currently not supported"?

![image](https://user-images.githubusercontent.com/1278803/83339965-a5496200-a298-11ea-8547-9c0afc6c9e8f.png)

**Comment by cartermp** [comment]
@dharmaturtle I don't see the error you're seeing on the latest build of the F# compiler.

![image](https://user-images.githubusercontent.com/6309070/83702290-24a0a380-a5c1-11ea-88be-e5c24fa0b584.png)
![image](https://user-images.githubusercontent.com/6309070/83702312-2f5b3880-a5c1-11ea-9524-af23a2daeb86.png)

**Comment by dsyme** [comment]
I listed my preferred resolution of the design items here: https://github.com/fsharp/fslang-design/issues/48#issuecomment-540387514

I've been through those comments above and I don't really see any reason to change anything in these resolutions.   As far as I can see that just leaves this one issue to be addressed:

> Issue Using nameof in a literal, e.g.., let [<Literal>] Foo = nameof Bar
> 
> Proposed resolution: Fix this 

I would also like be happy to see the issue with generic parameters resolved.  Since `nameof` is treated as do special it is possible that could be done via the syntax `nameof<'T>`.  But for me it is not critical to address this to activate the feature.

For me, I still fundamentally feel the other design issues just make `nameof` diverge more and more from standard name resolution, which means there is more cognitive overload in learning to use the feature, and more "ugh" feeling for me.  For example, allowing `nameof(M)` for an instance member just makes me confused about whether qualification is needed in F# when accessing instance members, and in what scope those names are active.

I also would really want `nameof` to align with any future features like `statichandleof` or `methodhandleof` and so on.  Those require precise resolution.

So that's where I stand on the design issues right now.

**Comment by TIHan** [comment]
@dsyme
At the moment, it feels like we are still looking at `nameof` with a reflection lens because of the `statichandleof` + `methodhandleof` mentions, and the by-design acceptance of getting compiled names of operators. In C#, `nameof` is a source code thing, not a reflection thing; meaning it's getting the name of the ident of what is in the source code. If we are going to implement our own `nameof`, we should follow similar principles but with a F# perspective; otherwise it would be confusing.

**Comment by dsyme** [comment]
> At the moment, it feels like we are still looking at nameof with a reflection lens because of the statichandleof + methodhandleof mentions, and the by-design acceptance of getting compiled names of operators.

Some key design principles are these:

1. To whatever extent possible, `nameof(M)` must be robust under rename-refactor of `M`.  That's kind of the point.  If you do a rename and `M` no longer resolves to what it used to resolve to then you have a problem.

2. Where possible, as little new logic must be added to the design of F#. It is better from both cognitive (ease-of-learning) and implementation perspectives to minimise the new logic and rules.

3. The feature should be simple and easy to use

For (1), to implement this accurately, we **must** resolve `M` to a symbol - and furthermore it must be the symbol the user expects.  To me that 100% means `M` must resolve to the same symbol that it would resolve to if it was used in a normal way at that point in the source code, using the same logic.  There may be some small differences to allow namespaces and generic parameters to be resolved, but that's all. This means we must do static resolution to a symbol.

For the question of "do we need to resolve to a specific overload - C# doesn't" - my answer is that for accurate renaming, then yes we do, and it is good that we do.  In F# tooling, it is possible to refactor-rename a specific overload - it's not the case that the whole overloaded method group is renamed.  For `nameof` to make sense then we must know if we should rename that symbol too or not.  I don't really mind what C# does here - because I'm looking at it from the perspective of what information F# needs in both language and tooling in order to do accurate F# renaming.

For the question of "why do we need to qualify instance members, C# doesn't" - my answer is that's becayse C# has unqualified access to instance member notation, but F# doesn't.  In F# there is literally no way to access an instance member except via dot notation.  And I really, really don't want to contemplate adding a whole lot of logic to allow that just for this feature. The testing space is huge for that kind of logic and there are a vast number of corner cases in it.

>  In C#, nameof is a source code thing, not a reflection thing; meaning it's getting the name of the ident of what is in the source code. 

I don't think this is really the case if the goal is accurate renaming. More specifically, I'd say the C# version is full of weird corner cases where symbols don't rename accurately.  This would be even more the case if you imagined a version of C# with the amount of shadowing that F# has added to the language.  If it's just a source thing and there are multiple ambiguous names for `M` then when do we rename and when don't we?

I really think we don't want the C# approach where `nameof(M)` is *not* accurate about what `M` refers to.  As an example, earlier we saw the suggestion to allow out-of-scope references like these:

```
let f x = ... nameof(f)... // f not in scpe
```
But immediately consider what happens in renaming one of the 'f' in the following:
```
let g() = 
    let f x = ... nameof(f)... // f not in scope? or is it? which f?
    let f x = ... nameof(f)... // f in scope? but which f?
    let f x = ... nameof(f)... // f in scope? but which f?
    f 3
let f() = g()
```
Choose a arbitrary 'f' and rename it and tell me what the resulting text will be.   

The point is: if you approach this with inaccurate/syntactic resolution, then in F# you will hit many many corner cases where renaming is inaccurate, in this case because there is so much shadowing around.

For this reason, there is the bsaic design principle that `nameof(thing)` has enough information to accurately resolve `thing` to a symbol.

> If we are going to implement our own nameof, we should follow similar principles but with a F# perspective; otherwise it would be confusing.

The C# syntactic design half-achieves certain goals. But putting it directly into the F# language design doesn't take into account how that car-crashes with aspects of the F# language design such as 
1. differences in scoping 
2. shadowing
3. the absence of unqualified instance member access in the language

**Comment by TIHan** [comment]
@dsyme I agree with everything here. I think we want the same thing. I just didn't communicate clearly what I meant by "reflection thing".

The example case of an operator -> `nameof (+)` would result in `op_Addition` because that name is what is actually compiled down; this is what I meant by "reflection thing". You don't see `op_Addition` in the source code, but it is part of the syntax tree.

Before I was thinking in terms of how C# was handling the cases for instance members, but now I am totally good with following the current scoping / name resolution rules when using `nameof`. Meaning: `nameof (Unchecked.defaultof<Vector3>.X)` is how you would get the field name. We do the same thing with code quotations: `| SpecificCall <@ Unchecked.defaultof<Vector3>.X @>`. It's very predictable, unlike `Vector3.X`.

So, I think I my two main concerns are:

- Using `nameof` on operators, do we really need it? Does it make sense?
- Using `nameof` on type arguments.

**Comment by abelbraaksma** [comment]
> Using nameof on operators, do we really need it? Does it make sense?

Yes, please! Though I would expect that both `nameof (+)` and `nameof X.op_Addition` would work. These things help in logging, and we can find out what's the actual compiled name this way, which is a great (albeit niche) addition.

> Using nameof on type arguments.

It'd be nice to have, I think, but for me, I can't really think of a use case, other than being rather surprised that it doesn't work by the time this goes into RTM releases.

**Comment by dsyme** [comment]
@TIHan thanks, that's great.  There is indeed a difference in the C# and F# design of these features - F# is aiming for a precise symbol resolution like in quotations and regular code.  I'll adjust the RFC to capture the design principles mentioned above.

> Using nameof on type arguments.

I am going to try to fix this this afternoon (if you don't already have a fix?)

**Comment by dsyme** [comment]
> Using nameof on operators, do we really need it? Does it make sense?

I think we can't avoid it on `nameof(X.op_Addition)` or `nameof(op_Addition)`.

I could see the case for it not being legal for `nameof(+)` however I think that's actually a really useful feature - I've very often had to go look up the naming for .NET and F# operators e.g. `++>` or even `op_Multiply` v. `op_Multiplication` where I've made mistakes.  

So I'd use `nameof(op)` quite often in real code I think.

Note you can call `nameof(++>)` etc. without the operator actually being in scope because of F#'s implicit operator expressions.

```
> nameof(++>);;
val it : string = "op_PlusPlusGreater"
```

**Comment by dsyme** [comment]
Here's the proposed resolution of the design issues: https://github.com/fsharp/fslang-design/pull/473

The last thing is allowing `nameof<'T>` for generic parameters.  I'll look at that now.

I also noticed `nameof(Unchecked.defaultof<C>.Property)` didn't work, I want to look at that again

As indicated in the text in #473 I do feel a slight discomfort at the resolution of the issue for operators - in other cases as shown we don't reveal the compiled name of things.  However 

1. operators really are treated in a very shallow way in F#.   

2. I checked and operator renaming does work as expected when `nameof(++>)` is used for example.  Tooling doesn't really support renaming on `op_Addition` - it's either one or the other.  

3. In most error reporting and diagnostics tools already report the compiled names for operators.  So it's already fairly natural for an error mesage to contain `op_Addition` even if that never actually occurs in the source code.

So I think having `nameof(+)` resolve to `op_Addition` is ok.

**Comment by dsyme** [comment]
I implemented two design changes here: https://github.com/dotnet/fsharp/pull/9498

1. Allow nameof<'T> for name of generic parameters

2. Allow `nameof(Unchecked.defaultof<C>.Property)` for instance members

I checked and literals are already allowed:

```
> let [<Literal>] x = nameof(id);;
val x : string = "id"

> x;;
val it : string = "id"
```

I think this resolves all issues in the nameof design but @TIHan and @cartermp will check

May require fixing up some tests

**Comment by dsyme** [comment]
We discussed the issue around operators and decided that the return of the compiled name is wrong

So we want:

    name(op_Addition) --> "op_Addition"
    name(+) --> "+"
    name(FSharp.Core.Operators.(+)) --> "+"
    name(FSharp.Core.Operators.op_Addition) --> "op_Addition"


There may be another feature `compilednameof(...)` that would give ```List`1```, `op_Addition` and so on

**Comment by abelbraaksma** [comment]
> We discussed the issue around operators and decided that the return of the compiled name is wrong

That's a pity, I kinda liked getting `op_Addition` etc, plus that it could be used in introspection through reflection. 

I've two follow up questions on the new design decisions:

* do we take the `CompiledName` or the F# name? The choice may matter in reflection cases.
* for instance members, is the creation of a throwaway object or value erased? It may not matter much for classes, but for value types you (probably) wouldn't want to have the overhead in the stack.

In the design it's said that the resolution is 100% compile time, leading to a literal with zero overhead, so I'd assume that whatever is in the expression gets erased.

**Comment by jwosty** [comment]
I think that ``nameof(expr)`` should definitely give the name as it appears in F# source code, and that there should be a future separate ``compilednameof(expr)``. If ``nameof`` were to sometimes give the compiled name, that would back us into a design corner, where what if I want a use case where I really do want to get back "+" (from the operator) or "iter" (from one of the F# modules)? If both existed, the difference would be intuitive

**Comment by jwosty** [comment]
Also, since ``nameof`` can be used in literals, shouldn't it also work as a pattern itself?

```
match x with
| nameof(foo) -> ...
| nameof(bar) -> ...
```

**Comment by TIHan** [comment]
@jwosty I think we have a PR that would enable that: https://github.com/dotnet/fsharp/pull/8754

@dsyme What do you think of that?

**Comment by dsyme** [comment]
> do we take the CompiledName or the F# name? The choice may matter in reflection cases.

The name occuring in the source

> for instance members, is the creation of a throwaway object or value erased? It may not matter much for classes, but for value types you (probably) wouldn't want to have the overhead in the stack

The code inside the `nameof(expr)` is not executed.  The expression is always replaced byu a string literal, including in quotations

**Comment by dsyme** [comment]
> do we take the CompiledName or the F# name? The choice may matter in reflection cases.

The name occuring in the source

> for instance members, is the creation of a throwaway object or value erased? It may not matter much for classes, but for value types you (probably) wouldn't want to have the overhead in the stack

The code inside the `nameof(expr)` is not executed.  The expression is always replaced byu a string literal, including in quotations

**Comment by abelbraaksma** [comment]
Thanks @dsyme, at clears things up. @jwosty, you and a compelling point, maybe we should already start I om a lang suggestion for `compilednameof`, while our memory is still fresh (in my own situation, I see a stronger need for the latter, but that's just me).

**Comment by dsyme** [comment]
So it turns out that 

    let rec [<Literal>] x = nameof(x)

is *way* harder than I had considered.

OK, so at the moment F# doesn't support literal definitions in recurisve scopes that reference other literals in the recursive scope recursively or not "in order left to right".  For example

```fsharp
let rec [<Literal>] x = 1
and y = x + 1
```

is allowed but

```fsharp
let rec [<Literal>] x = y+1
and y = 1
```

is not. This also applies to `namespace rec` and `module rec`, e.g. 
```fsharp
module rec SomeCode =
    let [<Literal>] x = 1
    let y = x + 1
```
The reason for this is that supporting this requires adding an additional phase to the processing of recursive definitions where literals are computed and commited into the symbols.  It also requires determining cycles in the graph of literal definitions.  We don't do either of these thigns at the moment.

This means that

```fsharp
let rec [<Literal>] x = 1
and y = nameof(x)
```
is allowed but
```fsharp
let rec [<Literal>] x = nameof(x)
```
is not.

I think we should consider this an orthogonal limitations since, as I show above, it already applies to other literal definitions.  There is not much likely use for `let rec [<Literal>] x = nameof(x)` (though I guess it might be away to define self-describing symbols without any actual string in your code, but I'm ok with not supporting that)

**Comment by TheJayMann** [comment]
I am okay with not supporting self defined constants, as it is not too difficult to work around, but would like to add that I do often use this pattern in C#.
```csharp
const string MagicText = nameof(MagicText);
```
Works well for cases where a hard coded literal text is justified, the literal text is well self documenting, and it is being used in multiple locations.  I've often used this pattern for the specifying the keys of an ASP.NET MVC TempData dictionary.

Of course, if it is not simple enough to justify the additional feature, it is simple enough to write `let rec [<Literal>] MagicString = nameof(MagicString)` as `let [<Literal>] MagicString = "MagicString"`.

**Comment by dsyme** [comment]
Well you can just do 

    let rec MagicString = nameof(MagicString)

or

module rec MagicStrings =
    let Magic = nameof(Magic)

You just don't get to define it as `[<Literal>]`.  

We can definitely consider lifting the restriction, but it should just be dealt with orthogonally I think.

**Comment by cartermp** [comment]
The use for something like `let rec [<Literal>] x = nameof(x)` comes up a bit in the F# tooling for VS, and might show up in other tooling contexts where something expects a "tag" of sorts that is ultimately a string value. It's nice not to have to type out the string literals. Obviously you can do `let rec x = nameof x` to accomplish a similar gaol, but not a literal.

Also verifying from earlier in the thread that this works:

```fsharp
type X() as this =
    inherit Dictionary<string, string>()
    member val XValue = this.TryGetValue(nameof this.XValue) |> function true, x -> ValueSome x | false, _ -> ValueNone
    member val YValue = this.TryGetValue(nameof this.YValue) |> function true, y -> ValueSome y | false, _ -> ValueNone
    member val ZValue = this.TryGetValue(nameof this.ZValue) |> function true, z -> ValueSome z | false, _ -> ValueNone
```

**Comment by cartermp** [comment]
Should I file an issue on dotnet/fsharp for the general problem with literals? Or is that a language suggestion?

**Comment by dsyme** [comment]
> Should I file an issue on dotnet/fsharp for the general problem with literals? Or is that a language suggestion?

I guess it's a language suggestion.  It's pre-approved however.

**Comment by dsyme** [comment]
Note that the topic overlaps with various suggestions for extending the range of guaranteed compile-time evaluation of expressions in F#.

.NET literals are frustrating because they are so limited - and languages don't really have the option of augmenting what's allowed.

**Comment by cartermp** [comment]
Added https://github.com/fsharp/fslang-suggestions/issues/889  - some of the example code in this thread actually requires another language suggestion, making your statement about overlap with other feature areas very true.
